torch/distributed/_checkpoint/resharding.py

```diff
@@ -0,0 +1,197 @@
+        storage_key = f"{storage_key_prefix}_{suffix}"
+
+        shard_size = 1
+        for d in shard_md.shard_sizes:
+            shard_size *= d
+
+        # not particularly great
+        storage_size = shard_size * tensor.local_shards()[0].tensor.element_size()
+
+        one_smd = StorageMetadata(
+            shard_metadata=shard_md,
+            storage_key=storage_key,
+            length=storage_size,
+            offset=0,
+        )
+        smd.append(one_smd)
+
+    return ExtendedTensorMetadata(
+        tensor_metadata=tensor.metadata(),
+        storage_metadata=smd,
+    )
+
+
+def prepare_sharded_tensor_write(
+    sharded_tensor: ShardedTensor,
+    storage_key_prefix: str,
+) -> Tuple[List[TensorWriteRequest], ExtendedTensorMetadata]:
+    """
+    Prepare sharded tensor write.
+
+    Args:
+        sharded_tensor: The sharded tensor to persist.
+        storage_key_prefix: The identifier to be prepended to storage keys
+                            associated with the sharded tensor.
+
+    Returns:
+        Write requests for persisting the sharded tensor, and metadata
+        describing the persisted sharded tensor.
+    """
+    write_requests = []
+    for shard in sharded_tensor.local_shards():
+        # each shard has its own storage key.
+        # For most cases, the read is a recovery from a failure to the same sharding
+        # and does not need any resharding, write each shard as is is the most effective
+        suffix = "_".join([str(i) for i in shard.metadata.shard_offsets])
+        storage_key = f"{storage_key_prefix}_{suffix}"
+
+        tensor = shard.tensor.detach()
+
+        wr = TensorWriteRequest(
+            tensor=tensor,
+            storage_key=storage_key,
+        )
+        write_requests.append(wr)
+    return write_requests, _compute_sharded_tensor_md(
+        storage_key_prefix, sharded_tensor
+    )
+
+
+def prepare_sharded_tensor_read(
+    metadata: ExtendedTensorMetadata, sharded_tensor_out: ShardedTensor
+) -> List[TensorReadRequest]:
+    """
+    Prepare sharded tensor read.
+
+    Args:
+        metadata: Metadata describing the persisted sharded tensor. Normally,
+                  this is generated by func::`prepare_sharded_tensor_write`.
+        sharded_tensor_out: The dest sharded tensor.
+
+    Returns:
+        A list of class::`TensorReadRequest`. When fullfilled,
+        `sharded_tensor_out`'s local shards load from the persisted sharded
+        tensor.
+    """
+    read_reqs = []
+    for shard in sharded_tensor_out.local_shards():
+        # scan all mds looking for chunks
+        for storage_md in metadata.storage_metadata:
+            shard_md_from_storage = storage_md.shard_metadata
+            tensor = shard.tensor.detach()
+            assert shard_md_from_storage is not None
+            # this is a naive quadratic algo that can later be optimized
+            #   by sorting metadata and the shards md
+            # FIXME how do we handle offset > 0?
```

Comment:
Because there's no place to pass it down.
The idea was that this would be used by the storage layer when bundling multiple tensors in a single file and this would be the offset into that file.

I removed this as I don't believe it's the right design.
This information can easily be kept by the storage layer itself.